ql/src/java/org/apache/hadoop/hive/ql/parse/repl/metric/MetricCollector.java

```diff
@@ -0,0 +1,73 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.hadoop.hive.ql.parse.repl.metric;
+
+import org.apache.hadoop.hive.ql.parse.SemanticException;
+import org.apache.hadoop.hive.ql.parse.repl.metric.event.ReplicationMetric;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * MetricCollector.
+ * In memory collection of metrics
+ */
+public class MetricCollector {
+  private Logger LOG = LoggerFactory.getLogger(MetricCollector.class);
+  private Map<Long, ReplicationMetric> metricMap = new ConcurrentHashMap<>();
+  private final long maxSize;
+
+  private static volatile MetricCollector instance;
+
+  private MetricCollector(long maxSize){
+    this.maxSize = maxSize;
+  }
+
+  public static MetricCollector getInstance(long maxSize) {
+    if (instance == null) {
+      synchronized (MetricCollector.class) {
+        if (instance == null) {
+          instance = new MetricCollector(maxSize);
+        }
+      }
+    }
+    return instance;
+  }
+
+  public void addMetric(ReplicationMetric replicationMetric) throws SemanticException {
+    if (metricMap.size() > maxSize) {
+      throw new SemanticException("Metrics are not getting collected. ");
+    } else {
+      if (metricMap.size() > 0.8 * maxSize) { //soft limit
```

Comment:
Shouldn't it be just one time warning? It will keep on printing this for every entry from 80%-100% with same message.

Reply:
Yes it should be printed for every entry.

Reply:
Yes it should be printed for every entry.

Reply:
It will keep on printing this for every entry from 80%-100% with same message. Which will not help know if it at stuck at 80 or is >80 but <100. Can we have real percentage printed there.